### ISSUE TITLE
[FR] Ability to add annotations via GUI

### ISSUE THREAD

**Issue by abravalheri** [kubevirt-manager/kubevirt-manager#81]
Thank you very much for kubevirt-manager! It is a very nice piece of software and super useful.

Currently the GUI for creating a new VM allows adding arbitrary labels:

![image](https://github.com/user-attachments/assets/e29de737-2ac0-4e76-bcaf-829d8fa3a75e)

... but not annotations. Annotations are very useful in the context of multus and other CNI plugins for fine control over the network (and probably they also have many other uses) - examples: https://kubeovn.github.io/docs/stable/en/reference/annotations/?h=annotation, https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/docs/how-to-use.md.

It would be nice if users could add annotations via the GUI. Probably something like what is currently done for labels would be enough.

Another approach would also be to allow the user to edit/fine-tune the YAML automatically generated by the GUI before it is applied on the cluster...

**Comment by feitnomore** [comment]
Thanks @abravalheri . I'll work to get an Annotations tab implemented.
I had another version to put out, but I've found some problems on Kubevirt that made me focus on, and Kubevirt-Manager has been stuck. I'll see if I can get anything out soon.

**Comment by abravalheri** [comment]
Thank you very much! Those are very great news.

**Comment by feitnomore** [comment]
@abravalheri I haven't had a chance to work on annotations on VirtualMachine level... for Multus I already have something implemented on the Network screen, if you have Multus, you'll see the Network menu on the left as well as options to select network attachment definitions when creating the VM.
For annotations related to MetalLB, they are mostly related to services, and on Load Balancer screen, as well as on Cluster screen you have annotations to use on it. 
I'll leave this issue opened, as I really plan to work on this, but I was owing the community a new version, and had not had a chance to work on this for 1.4.1.

Thanks!

**Comment by abravalheri** [comment]
Thank you very much @feitnomore for the pointers! It turns out that kubevirt does not deal very well with "pre-annotating" the VM YAML with multus annotations... It creates its own based on the network configuration and does not know how to merge annotations given by the user.

For my use case what I had to do is to "edit" the VM generated by kubevirt-manager (before turning the VM on for the firest time) and change the "type" from "bridge" to something a bit more complicated ([documented here](https://kubevirt.io/user-guide/network/net_binding_plugins/macvtap/)):

```diff
-          - bridge: {}
-            name: net2
+          - {name: net2, binding: {name: macvtap}}
```

Now I will test the same approach to see if I can add a third multus network (currently kubevirt-manager has a maximum of 2 networks).

**Comment by feitnomore** [comment]
If that is the case (because at home I use multus with bridge), maybe you could share some NetworkAttachmentDefinitions and I can give a few shots over here to see if I can add "macvtap" to the list of options (today we have bridge and masquerade).

**Comment by abravalheri** [comment]
Sure (thank you very much for having a look on this).

To get the macvtap plugin it is a bit complicated to install[^1] ([docs](https://kubevirt.io/user-guide/network/net_binding_plugins/macvtap/)), but this is what I did:

  1. First, prepare all the necessary files:

  ```bash
  mkdir -p ~/provision/kubevirt/cluster-network-addons-operator
  cd /provision/kubevirt/cluster-network-addons-operator
  
  wget https://github.com/kubevirt/cluster-network-addons-operator/releases/download/v0.95.0/namespace.yaml
  wget https://github.com/kubevirt/cluster-network-addons-operator/releases/download/v0.95.0/network-addons-config.crd.yaml
  wget https://github.com/kubevirt/cluster-network-addons-operator/releases/download/v0.95.0/operator.yaml
  
  wget https://github.com/kubevirt/cluster-network-addons-operator/releases/download/v0.95.0/network-addons-config-example.cr.yaml -O config-to-apply.yaml
  ```
  
  Modify the file `config-to-apply.yaml` according to your needs, for example:
  
  ```yaml
  ---
  apiVersion: networkaddonsoperator.network.kubevirt.io/v1
  kind: NetworkAddonsConfig
  metadata:
    name: cluster
  spec:
    imagePullPolicy: IfNotPresent
    # kubeMacPool: {}
    # kubeSecondaryDNS: {}
    # kubevirtIpamController: {}
    # linuxBridge: {}
    macvtap: {}
    # multus: {}  # already installed
    # multusDynamicNetworks: {}
    # ovs: {}
  ```
  
2. Now we perform the installation:

  ```bash
  kubectl apply -f namespace.yaml
  kubectl apply -f network-addons-config.crd.yaml
  kubectl apply -f operator.yaml
    
  kubectl apply -f config-to-apply.yaml

  kubectl wait networkaddonsconfig cluster --for condition=Available
  kubectl get networkaddonsconfig cluster -o yaml  # check
  ```

  A new `macvatp` executable should appear when you run `ls /opt/cni/bin`

3. Finally we need to enable the feature gate:

  ```bash
  kubectl patch kubevirts -n kubevirt kubevirt --type=json -p='[{
    "op": "add",
    "path": "/spec/configuration/developerConfiguration/featureGates/-",
    "value": "NetworkBindingPlugins"
  }]'
  
  kubectl patch kubevirts -n kubevirt kubevirt --type=json -p='[{"op": "add", "path": "/spec/configuration/network",   "value": {
              "binding": {
                  "macvtap": {
                      "domainAttachmentType": "tap"
                  }
              }
          }}]'
  ```

4. Only then we need to create the `NetworkAttachmentDefinition`:

  ```bash
  cat <<EOF > eno8303-macvtap.yaml
  kind: NetworkAttachmentDefinition
  apiVersion: k8s.cni.cncf.io/v1
  metadata:
    name: host-network
    annotations:
      k8s.v1.cni.cncf.io/resourceName: macvtap.network.kubevirt.io/eno8303  #<--- This is the name of the interface where I am adding the macvtap.
  spec:
    config: |
      {
        "cniVersion": "0.3.1",
        "name": "host-network",
        "type": "macvtap",
        "mtu": 1500
      }
  EOF
  kubectl apply -f eno8303-macvtap.yaml
  ```

[^1]: Initially I just wanted to do something simple with `macvlan`, but currently that is a [limitation of kubevirt](https://kubevirt.io/user-guide/network/interfaces_and_networks/#invalid-cnis-for-secondary-networks)... So I had to resort to the much more complicated `macvtap`.

**Comment by abravalheri** [comment]
I am sorry, I am not very familiar with angular, so not in a good place to try to help out in the project. I will try to find some time to have a look in the codebase and see if I can contribute somehow.

**Comment by feitnomore** [comment]
So, today basically I have two main combinations:

if user selects `podNetwork`, this is how networks will look like:
```
- name: net1
  pod: {}
```
and this is how `interfaces` will look like:
```
- name: net1
  bridge: {}
```
or
```
- name: net1
  masquerade: {}
```
and if user selects `something else`, this is how network will look like:
```
- name: net1
  multus:
    networkName: bridgeNetworkName
```
An this is how `interface` will look like:
```
- name: net1
  bridge: {}
```
or
```
- name: net1
  masquerade: {}
```

So I'm assuming what we can do is add something like a macvtap option, or, identify the network type through NetworkAttachmentDefinition type value, and do something like this for network:
```
- name: net1
  multus:
    networkName: MacVtapNetworkName
```
and something like this for interface:
```
-  name: net1
   binding: 
     name: macvtap
```

Does it sound reasonable?

**Comment by abravalheri** [comment]
Thank you very much @feitnomore, this plan sounds great!

**Comment by feitnomore** [comment]
@abravalheri , I thought the logic on the code was different, I was wrong...
Anyway I've tried to adapt a solution here for you, where you can choose the network and you can choose between Masq/Bridge/Macvtap... would you test it? This is the image:
```
kubevirtmanager/kubevirt-manager:dev
```

Thanks

**Comment by abravalheri** [comment]
Thank you very much @feitnomore.

The way I tested this was by downloading the `bundled.yaml`, modifying it to use the image that you provided and applying it on top of the existing installation:

```console
$ kubectl apply -f kubevirt-manager-bundled.yaml
namespace/kubevirt-manager unchanged
serviceaccount/kubevirt-manager unchanged
customresourcedefinition.apiextensions.k8s.io/images.kubevirt-manager.io unchanged
clusterrole.rbac.authorization.k8s.io/kubevirt-manager unchanged
clusterrolebinding.rbac.authorization.k8s.io/kubevirt-manager unchanged
clusterrole.rbac.authorization.k8s.io/kubevirt-manager-kccm unchanged
clusterrolebinding.rbac.authorization.k8s.io/kubevirt-manager-kccm unchanged
clusterrole.rbac.authorization.k8s.io/kubevirt-manager-cas-workload unchanged
clusterrole.rbac.authorization.k8s.io/kubevirt-manager-cas-management unchanged
priorityclass.scheduling.k8s.io/vm-standard unchanged
priorityclass.scheduling.k8s.io/vm-preemptible unchanged
deployment.apps/kubevirt-manager configured  <-----------------------------------------------
service/kubevirt-manager unchanged
```

The `macvlan` option does not seem to show up for the NIC 2 network:

![image](https://github.com/user-attachments/assets/454eb11c-8be2-4016-99ce-68053961ea94)

**Comment by feitnomore** [comment]
My bad, there were some auto loading stuff there, try this one:
```
kubevirtmanager/kubevirt-manager:dev
```

Thanks

**Comment by abravalheri** [comment]
Hi @feitnomore, thank you very much. I run into the following problem:

![image](https://github.com/user-attachments/assets/8615c002-d251-4b95-a8ea-c6ec52f2acc7)

```
admission webhook "virtualmachine-validator.kubevirt.io" denied
the request: spec.template.spec.networks[1].name 'net2' not found.,
spec.template.spec.domain.devices.interfaces[1].name '' not found.,
Network interface name can only contain alphabetical characters, numbers, dashes (-) or underscores (_)
```

**Comment by feitnomore** [comment]
would you run this with Developer Console opened, and save the request/response for me to take a look?
Thanks again for being patient.

**Comment by abravalheri** [comment]
Does the following help?

[debug.txt](https://github.com/user-attachments/files/16948717/debug.txt)

**Comment by feitnomore** [comment]
Let me take a look and get back to you... thanks!

**Comment by feitnomore** [comment]
I've been busy these days, but'll take a look this week.

NOTE TO SELF:
```
{"apiVersion":"kubevirt.io/v1alpha3","kind":"VirtualMachine","metadata":{"name":"test-ubuntu-24.04lts","namespace":"parallel-wireless","labels":{"kubevirt.io/domain":"test-ubuntu-24.04lts","kubevirt-manager.io/managed":"true","fw.kubevirt-manager.io/vm-name":"test-ubuntu-24.04lts","k8s.v1.cni.cncf.io/networks":"host-network"}},"spec":{"running":false,"template":{"metadata":{"labels":{"kubevirt.io/domain":"test-ubuntu-24.04lts","kubevirt-manager.io/managed":"true","fw.kubevirt-manager.io/vm-name":"test-ubuntu-24.04lts","k8s.v1.cni.cncf.io/networks":"host-network"}},"spec":{"domain":{"devices":{"disks":[{"name":"disk1","disk":{}},{"name":"disk3","disk":{"bus":"virtio"}}],"interfaces":[{"name":"net1","bridge":{}},{"binding":{"name":"macvtap"}}],"networkInterfaceMultiqueue":true},"firmware":{"bootloader":{"bios":{}}}},"networks":[{"name":"net1","pod":{}},
,"volumes":[{"name":"disk1","dataVolume":{"name":"parallel-wireless-test-ubuntu-24.04lts-disk1"}},{"name":"disk3","cloudInitNoCloud":{"userData":"#cloud-config\nmanage_etc_hosts: true\nhostname: test-ubuntu-24.04lts\nssh_pwauth: true\n","networkData":"version: 1\nconfig:\n    - type: physical\n      name: enp1s0\n      subnets:\n      - type: dhcp\n    - type: nameserver\n      address:\n      - '8.8.8.8'\n      - '8.8.4.4'\n"}}],"nodeSelector":{"kubernetes.io/hostname":"yamper"},"priorityClassName":"vm-standard"}},"instancetype":{"kind":"VirtualMachineClusterInstancetype","name":"1gb-1vcpu"}}}
```

**Comment by feitnomore** [comment]
I don't have binding on my system, but please, try `kubevirtmanager/kubevirt-manager:dev003`. 
I'll wait on your feedback.

**Comment by abravalheri** [comment]
Thank you very much @feitnomore, I tested here and I think it works!

**Comment by feitnomore** [comment]
Good news @abravalheri . I'll leave this issue as testing until I merge the fix, probably on `1.4.2`.
Thanks!

**Comment by feitnomore** [comment]
@abravalheri , how's it behaving so far? is it good enough to merge into the stable release? Thanks

**Comment by abravalheri** [comment]
Hi @feitnomore sorry for the delayed response. To be honest last time I created a VM in kubevirt was while testing this PR, but for the monitoring part it seems OK.
